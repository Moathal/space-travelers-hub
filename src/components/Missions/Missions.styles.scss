table {
  margin-right: 6em;
  margin-left: 6em;
  border: 2px solid #ecedee;
  border-collapse: collapse;

  thead {
    font-size: 15px;
    border: 2px solid #ecedee;
    border-collapse: collapse;

    tr {
      border: 2px solid #ecedee;
      border-collapse: collapse;

      .mission-header {
        width: 15%;
      }

      .description-header {
        width: 45%;
      }

      .status-header {
        width: 10%;
      }

      .button-header {
        width: 10%;
      }
    }
  }

  tbody {
    font-size: 12px;

    tr,
    td {
      border: 2px solid #ecedee;
      border-collapse: collapse;
      padding: 5px;
    }

    .mission-name {
      font-weight: bold;
      font-size: 15px;
    }

    .member-status-active {
      background-color: green;
      border-radius: 3px;
      padding: 1px;
      color: white;
      font-size: 11px;
    }

    .member-status-not-active {
      background-color: gray;
      border-radius: 3px;
      padding: 1px;
      color: white;
      font-size: 11px;
    }

    .mission-button-join {
      font-size: 11px;
      background-color: transparent;
      border: 1px solid green;
      color: green;
    }

    .mission-button-leave {
      font-size: 11px;
      background-color: transparent;
      border: 1px solid red;
      color: red;
    }
  }

  tr:nth-child(even) {
    background-color: #f2f2f2;
  }
}
