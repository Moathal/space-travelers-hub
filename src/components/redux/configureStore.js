import { configureStore } from '@reduxjs/toolkit';
import missionsSlice from './missions';
import rocketsReducer from './rockets';

export default configureStore({
  reducer: {
    rockets: rocketsReducer,
    missions: missionsSlice,
  },
});
