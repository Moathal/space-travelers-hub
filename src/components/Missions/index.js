/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { loadMissions, toggleMission } from '../../redux/missions';
import './Missions.styles.scss';

const Missions = () => {
  const dispatch = useDispatch();

  const missions = useSelector((state) => state.missions.missions);

  useEffect(() => {
    if (missions.length === 0) {
      dispatch(loadMissions());
    }
  }, []);

  const renderTableRows = () => (
    <tbody>
      {
        missions.map((mission) => (
          <tr key={mission.id}>
            <td className="mission-name">
              {mission.mission_name}
            </td>
            <td>
              {mission.description}
            </td>
            <td className="member-status">
              <span className={`member-status-${mission.reserved ? 'active' : 'not-active'} `}>
                {
                  mission.reserved ? 'Active Member' : 'Not A Member'
                }
              </span>
            </td>
            <td>
              <div className="button-container">
                <button
                  type="button"
                  className={`mission-button-${mission.reserved ? 'leave' : 'join'} `}
                  onClick={() => dispatch(toggleMission(mission.id))}
                >
                  {
                    mission.reserved ? 'Leave Mission' : 'Join Mission'
                  }
                </button>
              </div>
            </td>
          </tr>
        ))
      }
    </tbody>
  );

  return (
    <table>
      <thead>
        <tr>
          <th className="mission-header">Mission</th>
          <th className="description-header">Description</th>
          <th className="status-header">Status</th>
          <th className="button-header"> </th>
        </tr>
      </thead>
      {renderTableRows()}
    </table>
  );
};

export default Missions;
