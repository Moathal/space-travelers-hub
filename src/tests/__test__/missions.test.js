import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from '../../components/redux/configureStore';
import Missions from '../../components/Missions';

it('renders Missions correctly', () => {
  const tree = renderer.create(<Provider store={configureStore}><Missions /></Provider>).toJSON();
  expect(tree).toMatchSnapshot();
});
