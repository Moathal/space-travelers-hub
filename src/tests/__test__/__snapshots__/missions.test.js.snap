// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Missions correctly 1`] = `
<table>
  <thead>
    <tr>
      <th
        className="mission-header"
      >
        Mission
      </th>
      <th
        className="description-header"
      >
        Description
      </th>
      <th
        className="status-header"
      >
        Status
      </th>
      <th
        className="button-header"
      >
         
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
